File: packages/sdk-kotlin/src/main/kotlin/dev/rovenue/sdk/internal/FunnelClaimBridge.kt
// FunnelClaimBridge.kt — multicast between the Rust FunnelClaimListener
// callback and Kotlin coroutine consumers.
//
// Rust core's registerFunnelClaimListener(listener) accepts exactly one
// FunnelClaimListener at a time (per FFI design). The FunnelClaimBridge
// is the single registered listener; it fans out every resolved claim
// into a MutableSharedFlow that arbitrary Kotlin code can `.collect {}`
// from.
//
// Mirrors ObserverBridge exactly — payload is FunnelClaimResult instead
// of ChangeEvent.

package dev.rovenue.sdk.internal

import dev.rovenue.sdk.generated.FunnelClaimListener
import dev.rovenue.sdk.generated.FunnelClaimResult
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow

/** Single registered FunnelClaimListener fanning resolved claims into a
 *  SharedFlow — mirrors ObserverBridge. */
internal class FunnelClaimBridge : FunnelClaimListener {
    private val _flow = MutableSharedFlow<FunnelClaimResult>(
        replay = 0,
        extraBufferCapacity = 16,
        onBufferOverflow = BufferOverflow.DROP_OLDEST,
    )

    /** Called by UniFFI on the Rust listener thread.
     *
     *  tryEmit is non-suspending and returns false only when the buffer is
     *  full and DROP_OLDEST drops one — we silently accept the drop here.
     *  Callers should re-query server state if they miss an event. */
    override fun onFunnelClaimResolved(result: FunnelClaimResult) {
        _flow.tryEmit(result)
    }

    val flow: SharedFlow<FunnelClaimResult> = _flow.asSharedFlow()
}

File: packages/sdk-swift/Sources/Rovenue/Internal/FunnelClaimBridge.swift
//  FunnelClaimBridge.swift — multiplexer between the Rust funnel-claim
//  callback and multiple Swift AsyncStream subscribers.
//
//  Rust core's `registerFunnelClaimListener(listener:)` accepts exactly one
//  FunnelClaimListener at a time. To support multiple Swift consumers, we
//  register exactly one FunnelClaimBridge with the core and let the bridge
//  fan-out results into a table of AsyncStream continuations.
//
//  Mirrors ObserverBridge exactly, with payload type FunnelClaimResult.

import Foundation

/// Single registered `FunnelClaimListener` that fans resolved claims out to
/// AsyncStream subscribers — mirrors ObserverBridge.
internal final class FunnelClaimBridge: FunnelClaimListener, @unchecked Sendable {
    private let lock = NSLock()
    private var continuations: [UUID: AsyncStream<FunnelClaimResult>.Continuation] = [:]

    func onFunnelClaimResolved(result: FunnelClaimResult) {
        lock.lock()
        let snapshot = continuations.values
        lock.unlock()
        for c in snapshot { c.yield(result) }
    }

    func subscribe() -> AsyncStream<FunnelClaimResult> {
        AsyncStream { continuation in
            let id = UUID()
            lock.lock(); continuations[id] = continuation; lock.unlock()
            continuation.onTermination = { [weak self] _ in
                guard let self else { return }
                self.lock.lock(); self.continuations.removeValue(forKey: id); self.lock.unlock()
            }
        }
    }

    /// Testing hook — count of live continuations.
    func liveCount() -> Int {
        lock.lock()
        defer { lock.unlock() }
        return continuations.count
    }

    /// Testing hook — finish all streams and clear the table.
    func finishAll() {
        lock.lock()
        let toFinish = Array(continuations.values)
        continuations.removeAll()
        lock.unlock()
        for c in toFinish { c.finish() }
    }
}
